Drag a tab to the same pinned tab will crash

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

1. Open two panels. Both panels are contained a same tab. One of the tab is pinned.
2. Drag the unpin tab to the pinned panel.

https://github.com/user-attachments/assets/af9dd060-c8f2-4899-b4c5-f2b8c6d10799


### Environment

Zed: v0.161.2 (Zed)
OS: macOS 15.1.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2024-11-20T10:21:43.7013+08:00 [ERROR] {
  "thread": "main",
  "payload": "`at` split index (is 2) should be <= len (is 1)",
  "location_data": {
    "file": "/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/mod.rs",
    "line": 2319
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create",
    "backtrace::capture::Backtrace::new",
    "zed::reliability::init_panic_hook::{{closure}}",
    "std::panicking::rust_panic_with_hook",
    "std::panicking::begin_panic_handler::{{closure}}",
    "std::sys::backtrace::__rust_end_short_backtrace",
    "_rust_begin_unwind",
    "core::panicking::panic_fmt",
    "alloc::vec::Vec<T,A>::split_off::assert_failed",
    "workspace::pane::Pane::render_tab_bar",
    "<workspace::pane::Pane as gpui::element::Render>::render",
    "gpui::view::any_view::render",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::{{closure}}",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<workspace::pane_group::element::PaneAxisElement as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::window::WindowContext::draw",
    "gpui::window::Window::new::{{closure}}::{{closure}}",
    "gpui::window::Window::new::{{closure}}",
    "gpui::platform::mac::window::step",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run",
    "gpui::app::App::run",
    "zed::main",
    "std::sys::backtrace::__rust_begin_short_backtrace",
    "std::rt::lang_start::{{closure}}",
    "std::rt::lang_start_internal",
    "_main"
  ],
  "app_version": "0.161.2",
  "release_channel": "Zed",
  "os_name": "macOS",
  "os_version": "15.1.0",
  "architecture": "aarch64",
  "panicked_on": 1732069303701,
  "system_id": "fa4528a0-b7c9-4294-b316-b525fea6c724",
  "installation_id": "c89ec826-e61f-429d-9e13-ade2ab7b04ce",
  "session_id": "f38d2da4-c127-4a2a-add8-eb5499ac58d5"
}
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/workspace/src/pane.rs
+++ b/crates/workspace/src/pane.rs
@@ -2455,6 +2455,8 @@ impl Pane {
                         to_pane = workspace.split_pane(to_pane, split_direction, cx);
                     }
                     let old_ix = from_pane.read(cx).index_for_item_id(item_id);
+                    let old_len = to_pane.read(cx).items.len();
+                    move_item(&from_pane, &to_pane, item_id, ix, cx);
                     if to_pane == from_pane {
                         if let Some(old_index) = old_ix {
                             to_pane.update(cx, |this, _| {
@@ -2472,7 +2474,10 @@ impl Pane {
                         }
                     } else {
                         to_pane.update(cx, |this, _| {
-                            if this.has_pinned_tabs() && ix < this.pinned_tab_count {
+                            if this.items.len() > old_len // Did we not deduplicate on drag?
+                                && this.has_pinned_tabs()
+                                && ix < this.pinned_tab_count
+                            {
                                 this.pinned_tab_count += 1;
                             }
                         });
@@ -2484,7 +2489,6 @@ impl Pane {
                             }
                         })
                     }
-                    move_item(&from_pane, &to_pane, item_id, ix, cx);
                 });
             })
             .log_err();

